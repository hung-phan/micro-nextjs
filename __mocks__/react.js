// https://github.com/airbnb/enzyme/issues/1875
// This is to fix the error with React.memo

module.exports = {
  ...require("react"),
  memo: x => x
};
