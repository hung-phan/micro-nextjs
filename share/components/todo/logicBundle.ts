import update from "immutability-helper";
import { DefaultRootState } from "react-redux";
import actionCreatorFactory from "typescript-fsa";
import { reducerWithInitialState } from "typescript-fsa-reducers";
import { asyncFactory, thunkToAction } from "typescript-fsa-redux-thunk";

import { TodoModel } from "../../domain/model";
import fetch from "../../library/fetch";

export type State = TodoModel.Todo[];

export const mountPoint = "todoState";

export const selectors = {
  getCurrentState: (state: DefaultRootState): State => state[mountPoint],

  _findById: (internalState: State, id: string) =>
    internalState.find((todo) => todo.id === id),

  _findIndexById: (internalState: State, id: string) =>
    internalState.findIndex((todo) => todo.id === id),
};

const actionCreator = actionCreatorFactory(mountPoint);

const asyncActionCreator = asyncFactory<DefaultRootState>(actionCreator);

export const actions = {
  create: asyncActionCreator<string, TodoModel.Todo>("CREATE", (text) =>
    fetch("/api/todo", {
      method: "POST",
      body: JSON.stringify({ text }),
    }).then((res) => res.json())
  ),

  remove: asyncActionCreator<string, void>(
    "REMOVE",
    async (id, _dispatch, getState) => {
      const todo = selectors._findById(
        selectors.getCurrentState(getState()),
        id
      );

      if (!todo) {
        return;
      }

      try {
        await fetch(`/api/todo/${id}`, { method: "DELETE" });
      } catch (e) {
        // do nothing
      }
    }
  ),

  complete: asyncActionCreator<string, TodoModel.Todo>(
    "COMPLETE",
    async (id, _dispatch, getState) => {
      const todo = selectors._findById(
        selectors.getCurrentState(getState()),
        id
      );

      if (!todo) {
        return;
      }

      const result = await fetch(`/api/todo/${id}`, {
        method: "PUT",
        body: JSON.stringify({
          complete: !todo.complete,
        }),
      });

      return result.json();
    }
  ),

  fetch: asyncActionCreator<void, State>("FETCH", () =>
    fetch("/api/todo")
      .then((res) => res.json())
      .then((todoData: TodoModel.ITodo[]) =>
        todoData.map((opts: TodoModel.ITodo) => new TodoModel.Todo(opts))
      )
  ),
};

export const bindActions = {
  create: thunkToAction(actions.create.action),
  remove: thunkToAction(actions.remove.action),
  complete: thunkToAction(actions.complete.action),
  fetch: thunkToAction(actions.fetch.action),
};

export const reducer = reducerWithInitialState<State>([])
  .case(actions.create.async.done, (state, action) =>
    update<State>(state, {
      $push: [action.result],
    })
  )
  .case(actions.remove.async.done, (state, action) => {
    const todoIndex = selectors._findIndexById(state, action.params);

    if (todoIndex === -1) {
      return state;
    }

    return update<State>(state, {
      $splice: [[todoIndex, 1]],
    });
  })
  .case(actions.complete.async.done, (state, action) => {
    const todoIndex = selectors._findIndexById(state, action.params);

    if (todoIndex === -1) {
      return state;
    }

    return update<State>(state, {
      $splice: [
        [todoIndex, 1],
        [todoIndex, 0, action.result],
      ],
    });
  })
  .case(actions.fetch.async.done, (_state, action) => action.result)
  .build();
