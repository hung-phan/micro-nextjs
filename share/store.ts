import { Context, HYDRATE, MakeStore, createWrapper } from "next-redux-wrapper";
import { DefaultRootState } from "react-redux";
import {
  AnyAction,
  Store,
  applyMiddleware,
  combineReducers,
  compose,
  createStore,
} from "redux";
import thunkMiddleware, { ThunkMiddleware } from "redux-thunk";

import {
  State as TodoState,
  mountPoint as todoMountPoint,
  reducer as todoReducer,
} from "../share/components/todo/logicBundle";

export const middlewares = [
  thunkMiddleware as ThunkMiddleware<DefaultRootState, AnyAction>,
];

export const enhancers = [];

if (process.env.ENVIRONMENT === "client") {
  if (process.env.NODE_ENV === "development") {
    const { createLogger } = require("redux-logger");

    middlewares.push(createLogger({ level: "info" }));
  }

  if ((window as any).__REDUX_DEVTOOLS_EXTENSION__) {
    enhancers.push((window as any).__REDUX_DEVTOOLS_EXTENSION__());
  }
}

export interface RootState {
  [todoMountPoint]: TodoState;
}

const applicationReducer = combineReducers({
  [todoMountPoint]: todoReducer,
});

export const makeStore: MakeStore<Store<RootState>> = (_: Context) =>
  createStore(
    (state: RootState, action: AnyAction) => {
      switch (action.type) {
        case HYDRATE:
          return { ...state, ...action.payload };
        default:
          return applicationReducer(state, action);
      }
    },
    {}, // preloadedState
    compose(applyMiddleware(...middlewares), ...enhancers)
  );

export const wrapper = createWrapper<Store<RootState>>(makeStore, {
  serializeState: JSON.stringify,
  deserializeState: JSON.parse,
});
