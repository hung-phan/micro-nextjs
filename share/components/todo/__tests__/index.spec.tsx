import { render } from "enzyme";
import * as React from "react";
import { Provider } from "react-redux";

import TodoList from "..";
import { makeStore } from "../../../store";
import { actions } from "../logicBundle";

describe("todo/TodoList", () => {
  let store;

  beforeEach(() => {
    store = makeStore(null);
    store.dispatch(
      actions.fetch.async.done({
        result: [
          { id: "1", text: "Todo 1", complete: false },
          { id: "2", text: "Todo 2", complete: false },
          { id: "3", text: "Todo 3", complete: false },
          { id: "4", text: "Todo 4", complete: false },
        ],
      })
    );
  });

  it("should render component", () => {
    expect(
      render(
        <Provider key="provider" store={store}>
          <TodoList />
        </Provider>
      )
    ).toMatchSnapshot();
  });
});
