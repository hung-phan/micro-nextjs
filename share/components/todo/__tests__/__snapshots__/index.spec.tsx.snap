// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo/TodoList should render component 1`] = `
<div
  class="jsx-a681a50b99b62f37 container"
>
  <div
    class="jsx-a681a50b99b62f37 row"
  >
    <div
      class="jsx-a681a50b99b62f37 col-md-12"
    >
      <h1
        class="jsx-a681a50b99b62f37"
      >
        Todo List
      </h1>
    </div>
    <div
      class="col-md-12"
    >
      <div
        class="form-inline"
      >
        <div
          class="form-group"
        >
          <input
            class="form-control"
            placeholder="Todo"
            type="text"
          />
        </div>
        <button
          class="btn btn-success"
          type="button"
        >
          Add Todo
        </button>
      </div>
    </div>
    <div
      class="jsx-a681a50b99b62f37 col-md-12"
    >
      <table
        class="jsx-a681a50b99b62f37 table"
      >
        <tbody
          class="jsx-a681a50b99b62f37"
        >
          <tr>
            <td>
              <span>
                1
              </span>
            </td>
            <td>
              <span>
                Todo 1
              </span>
            </td>
            <td>
              <button
                class="btn btn-xs btn-success"
                type="button"
              >
                <svg
                  class="bi bi-check"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M13.854 3.646a.5.5 0 010 .708l-7 7a.5.5 0 01-.708 0l-3.5-3.5a.5.5 0 11.708-.708L6.5 10.293l6.646-6.647a.5.5 0 01.708 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
            <td>
              <button
                class="btn btn-xs btn-danger"
                type="button"
              >
                <svg
                  class="bi bi-trash"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.5 5.5A.5.5 0 016 6v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm2.5 0a.5.5 0 01.5.5v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm3 .5a.5.5 0 00-1 0v6a.5.5 0 001 0V6z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M14.5 3a1 1 0 01-1 1H13v9a2 2 0 01-2 2H5a2 2 0 01-2-2V4h-.5a1 1 0 01-1-1V2a1 1 0 011-1H6a1 1 0 011-1h2a1 1 0 011 1h3.5a1 1 0 011 1v1zM4.118 4L4 4.059V13a1 1 0 001 1h6a1 1 0 001-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <span>
                2
              </span>
            </td>
            <td>
              <span>
                Todo 2
              </span>
            </td>
            <td>
              <button
                class="btn btn-xs btn-success"
                type="button"
              >
                <svg
                  class="bi bi-check"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M13.854 3.646a.5.5 0 010 .708l-7 7a.5.5 0 01-.708 0l-3.5-3.5a.5.5 0 11.708-.708L6.5 10.293l6.646-6.647a.5.5 0 01.708 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
            <td>
              <button
                class="btn btn-xs btn-danger"
                type="button"
              >
                <svg
                  class="bi bi-trash"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.5 5.5A.5.5 0 016 6v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm2.5 0a.5.5 0 01.5.5v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm3 .5a.5.5 0 00-1 0v6a.5.5 0 001 0V6z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M14.5 3a1 1 0 01-1 1H13v9a2 2 0 01-2 2H5a2 2 0 01-2-2V4h-.5a1 1 0 01-1-1V2a1 1 0 011-1H6a1 1 0 011-1h2a1 1 0 011 1h3.5a1 1 0 011 1v1zM4.118 4L4 4.059V13a1 1 0 001 1h6a1 1 0 001-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <span>
                3
              </span>
            </td>
            <td>
              <span>
                Todo 3
              </span>
            </td>
            <td>
              <button
                class="btn btn-xs btn-success"
                type="button"
              >
                <svg
                  class="bi bi-check"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M13.854 3.646a.5.5 0 010 .708l-7 7a.5.5 0 01-.708 0l-3.5-3.5a.5.5 0 11.708-.708L6.5 10.293l6.646-6.647a.5.5 0 01.708 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
            <td>
              <button
                class="btn btn-xs btn-danger"
                type="button"
              >
                <svg
                  class="bi bi-trash"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.5 5.5A.5.5 0 016 6v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm2.5 0a.5.5 0 01.5.5v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm3 .5a.5.5 0 00-1 0v6a.5.5 0 001 0V6z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M14.5 3a1 1 0 01-1 1H13v9a2 2 0 01-2 2H5a2 2 0 01-2-2V4h-.5a1 1 0 01-1-1V2a1 1 0 011-1H6a1 1 0 011-1h2a1 1 0 011 1h3.5a1 1 0 011 1v1zM4.118 4L4 4.059V13a1 1 0 001 1h6a1 1 0 001-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <span>
                4
              </span>
            </td>
            <td>
              <span>
                Todo 4
              </span>
            </td>
            <td>
              <button
                class="btn btn-xs btn-success"
                type="button"
              >
                <svg
                  class="bi bi-check"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M13.854 3.646a.5.5 0 010 .708l-7 7a.5.5 0 01-.708 0l-3.5-3.5a.5.5 0 11.708-.708L6.5 10.293l6.646-6.647a.5.5 0 01.708 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
            <td>
              <button
                class="btn btn-xs btn-danger"
                type="button"
              >
                <svg
                  class="bi bi-trash"
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5.5 5.5A.5.5 0 016 6v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm2.5 0a.5.5 0 01.5.5v6a.5.5 0 01-1 0V6a.5.5 0 01.5-.5zm3 .5a.5.5 0 00-1 0v6a.5.5 0 001 0V6z"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M14.5 3a1 1 0 01-1 1H13v9a2 2 0 01-2 2H5a2 2 0 01-2-2V4h-.5a1 1 0 01-1-1V2a1 1 0 011-1H6a1 1 0 011-1h2a1 1 0 011 1h3.5a1 1 0 011 1v1zM4.118 4L4 4.059V13a1 1 0 001 1h6a1 1 0 001-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="jsx-a681a50b99b62f37 col-md-12"
    >
      <a
        class="jsx-a681a50b99b62f37"
        href="/static-page"
      >
        Go to static page
      </a>
    </div>
  </div>
</div>
`;
